cmd/erasure-healfile.go

```diff
@@ -36,48 +50,86 @@ func (s ErasureStorage) HealFile(offlineDisks []StorageAPI, volume, path string,
 			f.Checksums[i] = checksums[i]
 		}
 	}
-	blocks := make([][]byte, len(s.disks))
+
+	// Scan part files in disk, block-by-block reconstruct it and
+	// write to stale disks.
 	chunksize := getChunkSize(blocksize, s.dataBlocks)
-	for offset := int64(0); offset < size; offset += blocksize {
-		if size < blocksize {
-			blocksize = size
+	var chunkOffset, blockOffset int64
+	for ; blockOffset < size; blockOffset += blocksize {
+		// last iteration may have less than blocksize data
+		// left, so chunksize needs to be recomputed.
+		if size < blockOffset+blocksize {
+			blocksize = size - blockOffset
 			chunksize = getChunkSize(blocksize, s.dataBlocks)
 		}
+
+		// read a chunk from each disk, until we have
+		// `s.dataBlocks` number of chunks set to non-nil in
+		// `blocks`
+		blocks := make([][]byte, len(s.disks))
+		var buffer []byte
 		numReads := 0
 		for i, disk := range s.disks {
-			if disk != OfflineDisk {
-				if blocks[i] == nil {
-					blocks[i] = make([]byte, chunksize)
-				}
-				blocks[i] = blocks[i][:chunksize]
-				if !verifiers[i].IsVerified() {
-					_, err = disk.ReadFileWithVerify(volume, path, offset, blocks[i], verifiers[i])
-				} else {
-					_, err = disk.ReadFile(volume, path, offset, blocks[i])
-				}
-				if err != nil {
-					blocks[i] = nil
-				} else {
-					numReads++
-				}
-				if numReads == s.dataBlocks { // we have enough data to reconstruct
-					break
-				}
+			if disk == OfflineDisk {
+				continue
+			}
+			// allocate buffer only when needed - when
+			// reads fail, the buffer can be reused
+			if int64(len(buffer)) != chunksize {
+				buffer = make([]byte, chunksize)
+			}
+			if !verifiers[i].IsVerified() {
+				_, err = disk.ReadFileWithVerify(volume, path,
+					chunkOffset, buffer, verifiers[i])
+			} else {
+				_, err = disk.ReadFile(volume, path,
+					chunkOffset, buffer)
+			}
+			if err != nil {
+				// LOG FIXME: add a conditional log
+				// for read failures, once per-disk
+				// per-function-invocation.
+				continue
+			}
+
+			// read was successful, so set the buffer as
+			// blocks[i], and reset buffer to nil to force
+			// allocation on next iteration
```

Comment:
> blocks[i], buffer = buffer, nil

blocks[i] = buffer
buffer = nil

doing it in two separate lines will be more readable because reading from and writing to `buffer` in the same line is little difficult to read.

Reply:
I disagree on this. Any potential reader is expected to know some golang syntax. This style is idiomatic in golang and is used for example, to swap two variables - it is used in our codebase like this (try running `git grep '.*,[^:]*=[^{,(]*,[^,)}]*$'` to find all double-assignment statements in minio).

In addition, (1) this line is concise (and perhaps even elegant), and (2) it has a clear explanatory comment just above. 

Writing it as a swap operation would also be correct and idiomatic, but less obviously understandable - `blocks[i], buffer = buffer, blocks[i]`.

If you feel strongly about this, please let me know the reason. Even if one more reviewer feels it is difficult to read because it is a multi-assignment statement, I am ok to change it.

Reply:
> Writing it as a swap operation would also be correct and idiomatic, but less obviously understandable - blocks[i], buffer = buffer, blocks[i].

Swapping somehow looks easy on the eyes. Maybe because it is a recognised pattern.

> blocks[i], buffer = buffer, nil

This is a little harder to perceive immediately. My subjective view, hence you can ignore my comment.